### PR TITLE
Add bearer token API keys and move authentication credentials into PerRPCCredentials

### PR DESCRIPTION
**Pull request by tminusplus**
## What was changed
This PR achieves several things:
1. Adds support for bearer token API keys in the format of`$prefix_$id_$secret`.
2. Moves all authentication credentials into PerRPCCredentials to help prevent accidental leakage of secrets in code and to enforce secure connections when they are used.
3. Refactors the `GetServerConnection` tests from running a `urfave app` to setting up a `cli.Context` which parses flags and then connects to a server via `bufconn`.
4. Adds an `--insecure` flag, hidden to discourage use, to allow sending authentication over insecure connections. This is important for users who are using proxies like envoy which then sets up a secure connection from `localhost`.

## Why?

To add support, and tests, for our beta API keys in a way which encourages secure usage by default.

## Checklist
N/A

## Test Plan

- [x] Added tests for `oauth`, `apikeys`, and `HMAC`.
- [x] Tested a tcld command with `oauth` creds.
- [x] Tested a tcld command with `apikey` creds.

Since HMAC / request signatures was punted, I did not test the functionality but instead moved it over verbatim into the `apikey` credential to keep the refactor from being destructive.

### CODE REVIEW

File: app/connection.go [hunk 1]
```diff
@@ -4,67 +4,93 @@ import (
...
+	return ctx, conn, nil
+}
+
+func defaultDialOptions(c *cli.Context, addr *url.URL) ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+
+	creds, err := newRPCCredential(c)
 	if err != nil {
-		return nil, nil, err
+		return []grpc.DialOption{}, nil
```
Comment (mastermanu) [inline app/connection.go:53]: is this going to be confusing behavior to the user? Why did we deviate from returning an error?

File: app/credentials/apikey/apikey.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	if !c.insecure {
+		// Ensure the API key, AKA bearer token, is sent over a secure connection - meaning TLS.
+		if err := credentials.CheckSecurityLevel(ri.AuthInfo, credentials.PrivacyAndIntegrity); err != nil {
+			return nil, fmt.Errorf("the connection's transport security level is too low for API keys: %v", err)
+		}
+	}
+
+	if c.enableHMAC {
```
Comment (mastermanu) [inline app/credentials/apikey/apikey.go:90]: this feels premature to check-in to the code base?

File: app/connection.go (hunk 1 above)
Comment (tminusplus) [reply app/connection.go:53]: Ah that is a mistake, it should be returning the error. Will fix.

File: app/credentials/apikey/apikey.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	// RequestDatetime
+	// ActionName
+	requestSignatureFormat = "%s\n%s\n%s\n%s"
+)
+
+type Credential struct {
+	ID         string
+	secret     string // secret kept private to prevent accidental access.
+	enableHMAC bool
+	insecure   bool
```
[the remark this answers is not in the document]
Comment (tminusplus) [reply app/credentials/apikey/apikey.go:39]: Will do!

File: app/credentials/apikey/apikey.go (hunk 2 above)
Comment (tminusplus) [reply app/credentials/apikey/apikey.go:90]: It was previously in the codebase, so I wanted to move it along to the PerRPCCredentials so this PR wasn't destructive. Then my thought was we can remove it in a future PR when desired.

File: app/credentials/oauth/oauth.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+func WithInsecure(insecure bool) Option {
+	return func(c *Credential) {
+		c.insecure = insecure
+	}
+}
+
+type Credential struct {
+	accessToken string // keep unexported to prevent accidental leakage of the token.
+	insecure    bool
```
[the remark this answers is not in the document]
Comment (tminusplus) [reply app/credentials/oauth/oauth.go:26]: Will do, I initially did it to follow `curl`'s standard arg for this and have it follow through the code but `allowInsecureTransport` is nicer.

**Review by mastermanu — DISMISSED** [review]

**Review by anekkanti — APPROVED** [review]

**Pull request merged**